### ISSUE TITLE
Tailwind v4 @source directive ignores subdirectories with .gitignore

### ISSUE THREAD

**Issue by jhh** [tailwindlabs/tailwindcss#15452]
**What version of Tailwind CSS are you using?**

v4.0.0-beta.8

**What build tool (or framework if it abstracts the build tool) are you using?**

None, using CLI

**What version of Node.js are you using?**

v20.18.1

**What browser are you using?**

N/A

**What operating system are you using?**

tested on macOS and Linux

**Reproduction URL**

https://github.com/jhh/tailwind-gitignore

**Describe your issue**

Adding a .gitignore file to a directory referenced by a @source directive in a Tailwind CSS source file causes the referenced directory to be ignored.

I discovered this because the uv Python packaging tool creates a `.gitignore` in the `.venv` directory it uses and I couldn't `@source` a template directory within it. Manually deleting `.venv/.gitigore` fixes the issue.

**Comment by e-krebs** [comment]
this is a documented behavior: https://tailwindcss.com/docs/v4-beta#automatic-source-detection

**Comment by jhh** [comment]
I went over that documentation pretty carefully while I was trying to figure out why I wasn't picking up template sources. I think this is still a bug.


Works with one `.gitignore`:

```
.
├── .gitignore # ignores ignored/ dir
├── base.css
├── ignored
│   └── template.html # works as documented, @source will pick up classes
├── package-lock.json
├── package.json
├── README.md
└── test.html
```

Does not work when second `gitignore` is present in ignored directory:
```
.
├── .gitignore  # ignores ignored/ dir
├── base.css
├── ignored
│   ├── .gitignore # adding this, ignores *
│   └── template.html # this will not be seen by @source
├── package-lock.json
├── package.json
├── README.md
└── test.html
```

**Comment by e-krebs** [comment]
according to the (limited) doc, in neither of those two cases should the template file classesnbe picked up, so that's probably where the bug is.

**Comment by gregsullivan** [comment]
I'm encountering a variation on this issue unrelated to the `@source` directive. Because I suspect it is the same fundamental issue with automatic source detection, I figured it would be more helpful to post here.

I use a local development environment powered by Vagrant. The Vagrant configuration is itself a Git repository, and each site is based on a template. This results in a directory structure like this:

```
Vagrant
├── .git
└── www
     ├── domain.test
     │    └── ...
     └── another-domain.test
          ├── .gitignore [contains `*`]
          └── public_html
               ├── .git
               ├── .gitignore [contains `node_modules`]
               ├── package.json
               ├── tailwind.css
               └── etc.
```

Each domain folder (domain.test, another-domain.test, etc.) has it's own `.gitignore` file, and each of those files has `*` as an entry. (This is meant to protect your work from the site provisioner, which, according to a comment in the file, would otherwise delete your work.)

The result is that automatic source detection fails completely in `public_html` when `public_html` is the CWD. (It also fails if the CWD is in any subfolder of `public_html`, regardless of depth.)

However, if the `.gitignore` file in the parent folder of the CWD is updated to `public_html` instead of `*`, it works again. (`public_html/*` still fails.)

As an added wrinkle, #15684 applies here, and with the `--watch` flag, automatic source detection fails on the initial build, but works on subsequent builds while watching. So if one file in `public_html` is modified, that file's classes will be added, even with `*` in the `.gitignore` one level above the CWD.

I have created a minimal reproduction of the above which I'd be happy to share. If you'd prefer I open a separate issue, I can do that as well.

Thanks very much!

**Comment by davegaeddert** [comment]
I'm having the same issue with Python/uv as @jhh. The use case is exactly what the example in the docs describes, but for Python. It just doesn't seem to work because of the .gitignore * at the top of .venv, which seems like a bug. If I delete that file and but still have .venv in my repo gitignore, it works just fine.

I don't have a good test case on hand but I'm guessing if you dropped a gitignore * into a node_modules directory, it would break the example too?

![Image](https://github.com/user-attachments/assets/5b9b6d41-2df3-45a8-a61a-eadf21f0b79b)

**Comment by davegaeddert** [comment]
I just tried the latest release and it seems to be working now. @jhh?

**Comment by jhh** [comment]
> I just tried the latest release and it seems to be working now. [@jhh](https://github.com/jhh)?

I updated to Tailwind v4.0.9 and still see the issue in my [test repo](https://github.com/jhh/tailwind-gitignore).

**Comment by davegaeddert** [comment]
Yep, you're right. I was working on a different repo and for some reason it didn't have the .gitignore in the .venv itself.

**Comment by davegaeddert** [comment]
Ok, so I think I have some tests for this (of course they don't run in CI — custom runner maybe?): https://github.com/davegaeddert/tailwindcss/commit/d35be22da2921127882769f2d21b44baaf2cf24b

Interestingly to me, using `.venv` in a root .gitignore is also different from using `.venv/**/*`. I understand these are technically not the same, but as far as how the thing works, I'm surprised they don't act the same. Anyway, I'm wondering how much of this is just dependent on the behavior of the [ignore crate](https://crates.io/crates/ignore) (which has several open issues that may be related).

I wonder about just adding an option to the `@source` directive to turn off all the gitignore rules for a specific source? That would trickle down into the `WalkerBuilder` and turn off the git rules. Something like:

```css
@source "path" ignore(none);
/*
or @source "path" gitignore(false);
or something...
*/
```

Then it would be an explicit behavior people are opting in to, and the other uses cases for using @source don't change? @RobinMalfait would that make more sense so we can still point to a directory as a source?

**Comment by davegaeddert** [comment]
Or it could be something that ultimately sets `builder.parents(false)` on `WalkBuilder`.

**Comment by jordankimsey** [comment]
I am facing a similar issue related to the @source directive with Vite based applications. I am using version 4.0.9. I have a custom tailwind design system that I am referencing in my CSS file. In our Next.js / PostCSS applications the @source directive works and all of the tailwind styles are rendered properly.  This is not the case for our React Vite based applications.

**Comment by njacob1001** [comment]
Same issue here

**Comment by brittneypostma** [comment]
> I am facing a similar issue related to the [@source](https://github.com/source) directive with Vite based applications. I am using version 4.0.9. I have a custom tailwind design system that I am referencing in my CSS file. In our Next.js / PostCSS applications the [@source](https://github.com/source) directive works and all of the tailwind styles are rendered properly. This is not the case for our React Vite based applications.

Same issue here with SvelteKit vite based application with a design system imported using the [@source](https://github.com/source) directive, no component styles show up.

**Comment by sailorbob134280** [comment]
I just encountered an interesting variation on this bug: The tailwindcss binary ignores any and all sources if it is itself gitignored. 

**Background**
In my project, I didn't want to assume folks already had the tailwindcss binary downloaded, so I added a little script to fetch the latest version to my build system (it basically just performs the installation locally in the build directory, then uses that). But when the binary is gitignored, either explicitly or by being in a directory that's ignored, none of the sources show up (even if they're not ignored). This is pretty annoying, and I don't think it's intended behavior.

**Comment by izziaraffaele** [comment]
> I am facing a similar issue related to the [@source](https://github.com/source) directive with Vite based applications. I am using version 4.0.9. I have a custom tailwind design system that I am referencing in my CSS file. In our Next.js / PostCSS applications the [@source](https://github.com/source) directive works and all of the tailwind styles are rendered properly. This is not the case for our React Vite based applications.

I have this issue too and I discovered a very strange behaviour.... If I put the `@source` directive in the custom tailwind design system library it works with Vite application but not with NextJS. If I put the `@source` directive in the NextJS/Vite application it works only with NextJS as mentioned by @jordankimsey.

So a temporary workaround would be to have the `@source` directive in the design system library itself AND in the Vite/NextJS application

**Comment by philipp-spiess** [comment]
@jhh Hey! With the next patch release on we'll treat `.venv` and `venv` the same as `node_modules` so your use case should work as expected!

**Comment by gdarko** [comment]
This is still a bug in 2026?

**Comment by GrimLink** [comment]
@philipp-spiess, this remains an issue: `@source` continues to ignore all files when a `.gitignore` uses an "Allow List" (negation) pattern.

At Hyvä, [we’ve documented a workaround for our users](https://docs.hyva.io/hyva-themes/faqs/tailwindcss.html#gitignore-conflict-with-tailwind-v4-source), but the behavior remains highly confusing because Tailwind fails silently. There are no errors or warnings indicating that the `.gitignore` patterns are overriding the `@source` directive.

It would be a significant improvement if Tailwind offered an option to bypass this behavior or, at the very least, surfaced a warning when files are excluded due to gitignore rules.

**Comment by philipp-spiess** [comment]
@GrimLink I see. Mind filing a new issue for this?

**Comment by GrimLink** [comment]
Yeah no problem.
